Property Based Tests

clark-lindsay opened the pull request:
## Description

Adds property-based tests with [StreamData](https://hexdocs.pm/stream_data/0.5.0/StreamData.html) to improve confidence in the test suite and make the library more reliable and easier to extend.

Most of the property based testing is condensed around a single `property` test that asserts that an arbitrary `{schematic, data}` tuple, where the data and the schematic are compatible and will successfully `unify/2`, satisfies the following pseudocode: `data |> unify |> dump == data` 

As this is a draft, i am open to ***any and all feedback***

Closes #8

clark-lindsay committed:
chore: add property-based tests for some schematics

WORK IN PROGRESS: may end up removing the tests that the new
property-based tests overshadow, which would do away with the awkward
naming convention that i started with.

additionally, i wrote the generators in the `test/support` directory
after the JSON spec because it seemed to match well with the data types
supported by `schematic` right now (w/ the exception of `tuple`), but
this may just be incidental

clark-lindsay committed:
chore: add property-based tests for tuple schematics

WORK IN PROGRESS: there are a few rough edges here, so i'll record my
current thoughts on them:
  - i'm still not sure how valuable a separate module for `StreamData`
    generators is, but right now it feels like it improves the
    organization of the test code
  - currently a `list/0` schematic can match a non-homogenous list,
    whereas any call to `list/1` must match the given schematic
    homogenously (but you **could** pass a `one_of` schematic...)
    - is this the right design? i think it ***is right***, but i left
      myself a "TODO" to think about it some more
  - should property-based tests be written to describe the  properties
    of the code for the failure paths? or is that better left to
    hand-written edge case unit tests?

clark-lindsay committed:
chore: add property-based tests for `oneof` schematics

clark-lindsay committed:
chore: add property-based tests for `map` unification

also extended some of the data generation in the `Generators` test
support module

mhanberg commented on test/schematic_test.exs:
```diff
@@ -56,14 +75,37 @@ defmodule SchematicTest do
       assert {:ok, input} == unify(schematic, input)
     end
 
+    # TODO: does a list have to be homogenous in type?
```
Technically yes, but the single schematic can be a oneof schematic.

mhanberg commented on test/schematic_test.exs:
```diff
@@ -105,6 +177,48 @@ defmodule SchematicTest do
       assert {:ok, %{"foo" => "hi there!"}} == unify(schematic, input)
     end
 
+    test "map/1 schematic options properties" do
```
Doesn't ExUnitProperties import a `property` macro?

clark-lindsay replied:
yeah! i had trouble with it though because using it imports everything from `StreamData` into the body which gave me naming conflicts with `Schematic` functions. i guess we could not `import Schematic` at the top of the test module, or we could just use fully qualified names in the few places where there are conflicts. 

if you think either of those would read better then i’m down to make the changes

mhanberg replied:
are you sure? I think that is what `use ExUnitProperties` does, and you are already calling that.

I think that the property macro is just the same as the test macro, but registers them as properties so you can do something like `mix test --only properties`

clark-lindsay replied:
yeah maybe i did something wrong, and then just moved on without double checking. i’ll try again 🙂 thanks for pointing it out, as i forgot to mention it in my description for the PR

clark-lindsay replied:
you were right on all counts! (except that the mix command is `mix test --only property`, in the singular). committed the change [here](https://github.com/clark-lindsay/schematic/commit/316f3c462875b1bc0330205778a672576a4b8738)

thanks again for pointing this out so that i could double back and make this improvement. not sure what mistake i made earlier that led me to believe that that was going to be more trouble than it was worth.

mhanberg reviewed (commented):
I'll have to pull this down and play around with it, nice work so far!

clark-lindsay committed:
chore: use `property` over `test` for better ergonomics

- better reporting from `mix test`
- easier to notice/find/grep-for tests are property-based
- automatically tags the tests, so one can use `mix test --only
  property` to target _just_ the property-based tests

clark-lindsay committed:
chore: extend `tuple/2` property test w/ `from: :list` option

clark-lindsay committed:
chore: add more property tests for `map/1`

also extend and start to document support generators, in support of
ongoing test expansion and improvements

mhanberg commented:
Does it make sense to organize the prop tests based on "properties" of the code rather than by function?

like, the property is that inputs can be unified and dumped back and forth, so the stream data generator generates random arbitrary inputs/schematics and ensures that `input == unify(schematic, input) |> then(&dump(schematic, &1))`

I assume it's currently broken up by function because you were just incrementally working through them, but before we merge it is worth considering whether combining them into fewer properties makes sense.

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,98 @@
+
+  import Schematic
+
+  def json_primitive() do
```
let's not do the json related thing.

schematic can be used for validating external json data, but it can also be used from elixir->elixir data structures, which means it should be tested with all beam terms.

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,98 @@
+    |> StreamData.one_of()
+  end
+
+  def from_schematic(%Schematic{kind: kind}) do
```
I think this is closer to what we want.

I think there should probably be some function like `gen` or `generate` that takes a schematic, and then spits out the stream data equivalent.

I think you still need to move the stream data generator for the blueprintable map schematic into this.

Then, I feel like you would be able to to more or less have a single property that generates an arbitrary schematic that can be any depth and combination of them, and test that it unifies and dumps back to itself.

mhanberg replied:
And this functionality can probably be re-used to make the user facing stream data API

clark-lindsay replied:
i just added something that is getting towards this idea. it's a little rough, and doesn't cover key transformation for maps, but with this is place we can probably drop a lot of the smaller `property` tests that i wrote along the way to this. i'll take a look at that tomorrow

mhanberg commented on test/schematic_test.exs:
```diff
@@ -105,6 +178,89 @@ defmodule SchematicTest do
+      check all(
+              {blueprint, input} <-
+                StreamData.bind(
+                  StreamData.map_of(StreamData.binary(), Generators.schematic(), min_length: 1),
```
map schematics can technically have any value as a key in the blueprint, so we would want to expand this generator to reflect that.

We'll also want it to generate map blue prints that reflect the key transformation capabilities: https://hexdocs.pm/schematic/Schematic.html#map/1-transforming-keys

clark-lindsay committed:
fix: `dump/2` should not drop empty-string keys

i don't entirely understand why there was a call to `Map.delete/2` in
the `unify` definition for the `map` schematic, but with it in place i
was dropping `""` keys when `dump`-ing.

removing the call has no effect on any other tests, and so i believe
that it's not necessary, ***but*** i may have written the test
incorrectly and misunderstood the desired behaviour...

clark-lindsay committed:
chore: generate tuple schematics as empty `{}`

i still can't think of a good way to generate tuples, but i at least
want the schematic `kind` accounted for in the tests in the meantime

clark-lindsay commented:
> Does it make sense to organize the prop tests based on "properties" of the code rather than by function?
> 
> like, the property is that inputs can be unified and dumped back and forth, so the stream data generator generates random arbitrary inputs/schematics and ensures that `input == unify(schematic, input) |> then(&dump(schematic, &1))`
> 
> I assume it's currently broken up by function because you were just incrementally working through them, but before we merge it is worth considering whether combining them into fewer properties makes sense.

yeah, i think that that makes sense. i didn't jump into this with a wholistic understanding of everything, so yeah i was just working through incrementally. now that i have a better understanding of the shape of the whole thing, i think this makes sense

clark-lindsay committed:
chore: add new generator and property-test for arbitrary schematics

clark-lindsay committed:
chore: remove unused generator

mhanberg commented on test/schematic_test.exs:
```diff
@@ -16,6 +19,13 @@ defmodule SchematicTest do
   end
 
   describe "unify" do
+    property "input |> unify |> dump == input" do
+      check all({schematic, input} <- Generators.schematic_and_data()) do
+        {:ok, input} ==
+          unify(schematic, input) |> then(fn {:ok, result} -> dump(schematic, result) end)
+      end
+    end
```
will this replace the other properties?

clark-lindsay replied:
most of them, i think. i'm gonna comb through today and remove the ones that are definitely overshadowed by this

mhanberg replied:
👏🏻

clark-lindsay replied:
replaced a bunch of existing properties that were superseded by this one. still a few remaining, but i suspect only one or two besides this are worth keeping. lmk if you feel differently

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,187 @@
+  @doc """
+  Generates structure-less data that has an analogous `Schematic` function. This excludes maps, lists, and tuples.
+  """
+  def leaf_value() do
```
`scalar` might be the right word

clark-lindsay replied:
AHHHH!! thank you. i was just staring at it and thought "eh whatever i'll think of something better later".
thank you.

clark-lindsay replied:
fixed this in a recent commit

clark-lindsay committed:
chore: rename generator function for scalar data

thanks @motch for the _much_ improved name idea

clark-lindsay committed:
chore: remove `property`s that have been superceded

the `property` "input |> unify |> dump == input" supercedes most of the
existing `property`s in the test file, which were written as stepping
stones for me to understand the behaviours and design of the code to
work towards more general properties that could be tested

remaining `property`s that were not deleted still have cases that are
not covered by the aforementioned "input |> unify |> dump == input"
test, and some `TODO`s have been added to address the generation of
schematics and data so that these `property`s will become redundant and
then be removed

clark-lindsay committed:
fix: `tuple/2` unification should error if input is wrong length

also:
  - extended the schematics that can be generated from
    `Generators.schematic_and_data/0`
  - removed a `property` that is now superseded by the "input |> unify
    |> dump |> input" `property` test
  - added a few specific cases to the `tuple/2` example-based tests that
    reproduced the bug found in the `unify` callback from `tuple/2`

mhanberg commented:
One annoying comment, can you add stream_data to the formatter.exs import deps and then remove the parens from all the `all` instances? 

😅

clark-lindsay commented:
> One annoying comment, can you add stream_data to the formatter.exs import deps and then remove the parens from all the `all` instances?
> 
> 😅

haha sure thing. not "annoying" btw, i totally get it

clark-lindsay committed:
chore: import `stream_data` formatting & fix up formatting

clark-lindsay committed:
chore: improve data generation options and documentation

clark-lindsay committed:
chore: rename fns, reorganize, and add docs for `Generators`

mhanberg commented:
@clark-lindsay Can you rebase this and then I'll review

mhanberg commented on lib/schematic.ex:
```diff
@@ -367,41 +367,46 @@ defmodule Schematic do
-              {:ok, output} ->
-                {:cont, {:ok, [output | acc]}}
+        cond do
+          condition.(input) and length.(input) != Enum.count(schematics) ->
```
It might be better to alter the original error message to be `message = fn -> "a #{from} of...` and then change the brackers from `[`/`]` to `{`/`}`

and then instead of the `cond`, just change the if to `condition.(input) and length.(input) == length(schematics)`

and then in the else, change it to `{:error, ~s|expected #{message.()}|}`

clark-lindsay replied:
yeah that makes sense to me. i'll make the change

mhanberg commented on test/schematic_test.exs:
```diff
@@ -63,25 +73,35 @@ defmodule SchematicTest do
 
     test "tuple/2" do
-      schematic = tuple([int(), str(), map(%{alice: any()})])
+      schematic = Schematic.tuple([int(), str(), map(%{alice: any()})])
```
are you able to change back the `Schematic.tuple` to `tuple` ?

clark-lindsay replied:
ah, yeah i should be. thanks for the catch

mhanberg commented on test/schematic_test.exs:
```diff
@@ -16,6 +19,13 @@ defmodule SchematicTest do
   describe "unify" do
+    property "input |> unify |> dump == input" do
+      check all {schematic, input} <- Generators.schematic_and_data() do
+        {:ok, input} ==
```
```suggestion
        assert {:ok, input} ==
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+  import Schematic
+
+  @doc """
+    Generates a `{schematic, data}` tuple where the former describes the latter.
```
this doc shouldn't be indented like it is

mhanberg replied:
```suggestion
    Generates a `{schematic, data}` tuple where the former specifies the latter.
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+  Generators for property-based testing.
+
+  Every public function in this module returns a `StreamData` generator stream that can be used
+  directly or composed with other functions from the `StreamData` API.
```
```suggestion
  `StreamData` generators.
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+
+    ## Options
+
+    * `:using` - a list of `Schematic.kind` values that determines the **only** data types that **should** be generated. See the `Schematic` module for available `kind`s. Defaults to `[]`.
```
```suggestion
    * `:using` - a list of `Schematic.kind` values that determines the data types that will be generated. See the `Schematic` module for available `kind`s. Defaults to `[]`.
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+
+    * `:using` - a list of `Schematic.kind` values that determines the **only** data types that **should** be generated. See the `Schematic` module for available `kind`s. Defaults to `[]`.
+       * e.g. `["null", "integer", "boolean"]`
+    * `:excluding` - a list of `Schematic.kind` values that determines the **only** data types that **should not** be generated. See the `Schematic` module for available `kind`s. This option is mutually exclusive with `:using`; if the `:using` option is **not** an empty list then this option is ignored. Defaults to `[]`. 
```
```suggestion
    * `:excluding` - a list of `Schematic.kind` values that determines the data types that will not be generated. See the `Schematic` module for available `kind`s. This option is mutually exclusive with `:using`. Defaults to `[]`. 
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+      case using do
+        [] -> {:reject, excluding}
+        _ -> {:accept, using}
+      end
```
i think you can merge this case and lines 49-52

```elixir
|> then(fn kinds ->
  case using do
    [] -> Enum.reject(kinds, &Enum.member?(excluding, &1))
    _ -> Enum.filter(kinds, &Enum.member?(using, &1))
  end
end)
```

I'd probably switch to an if tho

```elixir
if Enum.empty?(using) do
  # ...
else
  #...
end
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+      case using do
+        [] -> {:reject, excluding}
+        _ -> {:accept, using}
+      end
```
same deal as above

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+      "map" ->
+        StreamData.map_of(
+          StreamData.one_of([StreamData.binary(), StreamData.integer()]),
+          data(depth + 1),
```
actually, i would also remove one of these function clauses and change this line to. too much code duplication for the one line of change that halts the recursion
```suggestion
          case depth + 1 do
            8 -> scalar()
            depth -> data(depth)
          end,
```

clark-lindsay replied:
okay. pretty sure i understand what you're saying. i'll make the change

clark-lindsay replied:
it's _much_ better after those changes, thanks for the suggestion

mhanberg replied:
Excelente

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+    |> StreamData.one_of()
+  end
+
+  defp schematic_from_data(data) when is_integer(data),
+    do:
```
sort of a mitch-ism, but i heavily disklike these multi-line keyword style function clauses.

if the body is more than one line, always do `do`/`end`

clark-lindsay replied:
makes sense. each of these started as one-liners and then grew so i don't think i noticed that the shape of the function looked kinda wack

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+      Enum.reduce(data, [], fn datum, acc ->
+        [schematic_from_data(datum) |> Enum.fetch!(1) | acc]
+      end)
+      |> oneof()
```
because StreamData and schematic have a similar API, i wonder if we should avoid importing Schematic in this module and have everything be a remote call `Schematic.oneof()`

clark-lindsay replied:
i could go either way, but the latter is certainly more clear. i'll drop the import

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+    ])
+  end
+
+  # TODO: generate schematics that allow for optional keys, and maybe add extra optional keys that aren't present in `data
```
```suggestion
  # TODO: generate schematics that allow for optional keys, and maybe add extra optional keys that aren't present in `data`
```

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+  end
+
+  @doc """
+  Generates structure-less data that has an analogous `Schematic` function. This excludes maps, lists, and tuples.
```
i think the phrase would be `non-compound` rather than `structure-less`

mhanberg commented on test/support/generators.ex:
```diff
@@ -0,0 +1,293 @@
+  end
+
+  @doc """
+  Generator for random, simple `Schematic`s. All `Schematic`s that this is capable of generating are
```
an instance of a schematic should be referred to as a "schematic" and not a "Schematic"

clark-lindsay committed:
chore: update generator docs

Co-authored-by: Mitchell Hanberg <mitch@mitchellhanberg.com>

clark-lindsay committed:
chore: fix up docs & use remote calls over imports to avoid confusion

clark-lindsay committed:
chore: clean up `Generators` implementation

clark-lindsay committed:
fix: improve error messages and implementation of `tuple/2`

clark-lindsay committed:
fix: back-fill `then/2` for older elixir versions

clark-lindsay committed:
fix: ensure `:stream_data` is started for elixir versions < 1.11

mhanberg commented on lib/schematic.ex:
````diff
@@ -353,9 +353,9 @@ defmodule Schematic do
 
   ```elixir
-  iex> schematic = tuple([str(), int()])
+  iex> schematic = Schematic.tuple([str(), int()])
````
did you say you could remove the `Schematic.` part?

mhanberg committed:
chore: update doc

mhanberg commented:
Thanks!
